### PR TITLE
Fix `NaN` comparison in `torch.median`

### LINKED ISSUE #38018: median with float64, lot of nans, and dim=1 returns random results in CUDA
## 🐛 Bug

Not sure if it is related to #30818.

```
x = torch.full((20, 3000), np.nan, dtype=torch.float64).cuda()
x[:, :280] = 1.1
torch.median(x, dim=1).values
```

tensor([ 1.1000, 29.4100, 15.5200, 17.0700, 18.7800, 20.6600, 22.7300, 25.0000,
        27.5000, 27.0000, 15.5200, 17.0700, 18.7800, 20.6600, 22.7300, 25.0000,
        27.5000, 27.0000, 15.5600, 20.5500], device='cuda:0',
       dtype=torch.float64)

```
torch.median(x.to(torch.float32), dim=1).values
```
tensor([-1.5882e-23,  1.8875e+00, -2.0000e+00,  2.9595e+00,  2.7350e+00,
         2.7667e+00,  1.0842e-19,  2.7934e+00, -2.0000e+00,  2.8228e+00,
         2.0000e+00,  2.8552e+00,  2.8906e+00,  2.9297e+00,  2.9219e+00,
         2.8906e+00,  2.9297e+00,  2.9219e+00,  2.9219e+00,  2.9219e+00],
       device='cuda:0')

```
torch.median(x.cpu(), dim=1).values
```
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       dtype=torch.float64)

## Environment

PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 445.87
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0
[pip] torchvision==0.6.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              h74a9793_1
[conda] libblas                   3.8.0                    14_mkl    conda-forge
[conda] liblapack                 3.8.0                    14_mkl    conda-forge
[conda] mkl                       2019.4                      245
[conda] mkl-service               2.3.0            py37hb782905_0
[conda] mkl_fft                   1.0.15           py37h14836fe_0
[conda] mkl_random                1.1.0            py37h675688f_0
[conda] numpy                     1.18.1           py37h93ca92e_0
[conda] numpy-base                1.18.1           py37hc3f5095_1
[conda] pytorch                   1.5.0           py3.7_cuda102_cudnn7_0    pytorch
[conda] torchvision               0.6.0                py37_cu102    pytorch

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @ngimel

### PR DESCRIPTION
**Pull request by emcastillo**
Fixes #38018 

when calling `eq_with_nan(v, kValue)` having `v` and `kValue` both `nan` is returning `false` when it should be `true`.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/SortingKthValue.cu#L76

The implementation is using intrinsics such as `__double_as_longlong` and comparing their bit representations. But the values of the bits obtained for both nans are different.
`9221120237041090560` for `v`
`9223372036854775807` for `kValue`

two different nans have different bit representations, so we have to do additional comparisons to fix this.

I changed this comparison and it seems to be working now.
However, when compared to a CPU implementation, the returned indices for the values seems to be random but valid. 
Probably this is an effect of the comparison order in the Cuda version.
I am not sure if this is ok since all the indices point to valid elements.

For the snippet in the issue I get the following:

```
# CUDA Values
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       device='cuda:0', dtype=torch.float64)
# CUDA indices
tensor([304, 400, 400, 528, 304, 304, 528, 336, 304, 432, 400, 280, 280, 336,
        304, 336, 400, 304, 336, 560], device='cuda:0')
```
```
# CPU values
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       dtype=torch.float64)
# CPU indices
tensor([515, 515, 515, 515, 515, 515, 515, 515, 515, 515, 515, 515, 515, 515,
        515, 515, 515, 515, 515, 515])
```

Also, maybe its better to change the `eq_with_nan` implementations to address this instead?
I am not sure if this will cause code to break in other places though ...

### CODE REVIEW

File: aten/src/ATen/native/cuda/SortingKthValue.cu [hunk 1]
```diff
@@ -73,8 +73,9 @@ __global__ void gatherKthValue(
     bool inRange = (i < inputSliceSize);
     scalar_t v = inRange ? doLdg(&inputSliceStart[i * inputWithinSliceStride])
                          : static_cast<scalar_t>(0);
-    bool isKValue = inRange && THCNumerics<scalar_t>::eq_with_nan(v, kValue);
-
+    // Check for NaN values, comparing two different NaNs using eq_with_nan
+    // returns always false, even if both values are nan
+    bool isKValue = inRange && ( (v == kValue) || (v != v) && (kValue != kValue));
```
Comment (emcastillo) [inline aten/src/ATen/native/cuda/SortingKthValue.cu:78]:
```suggestion
    bool isKValue = inRange && ((v == kValue) || (v != v) && (kValue != kValue));
```

**Comment by ngimel** [comment]
Thanks for the fix @emcastillo! According to my grepping, eq_with_nan is used only in kthvalue, it is also not part of std:: api's that THCNumerics aims to emulate, so do you mind removing eq_with_nan from THCNumerics completely?

**Comment by ngimel** [comment]
Returning different (but valid) indices compared to CPU is acceptable, we don't make guarantees as to which index will be returned.

File: aten/src/ATen/native/cuda/SortingKthValue.cu [hunk 2]
```diff
@@ -73,8 +73,9 @@ __global__ void gatherKthValue(
     bool inRange = (i < inputSliceSize);
     scalar_t v = inRange ? doLdg(&inputSliceStart[i * inputWithinSliceStride])
                          : static_cast<scalar_t>(0);
-    bool isKValue = inRange && THCNumerics<scalar_t>::eq_with_nan(v, kValue);
-
+    // Check for NaN values, comparing two different NaNs using eq_with_nan
+    // returns always false, even if both values are nan
+    bool isKValue = inRange && ((v == kValue) || (v != v) && (kValue != kValue));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SortingKthValue.cu:78]: you can use THCNumerics<scalar_t>::isnan here.

**Comment by emcastillo** [comment]
Updated, PTAL! 🙂

File: test/test_torch.py [hunk 3]
```diff
@@ -8864,6 +8864,20 @@ def test_multidim(x, dim):
...
+            # check reducing median with NaNs
+            # If the element in the median is a NaN, there can be issues
+            # when comparining with other nan elements
+            if fn_name == 'median':
+                y = torch.full((1, 3), np.nan, dtype=torch.float64, device=device)
+                y[:, :1] = 1.1
+                values, indices = fn_tuple(y, dim=1)
+                expected_values = torch.tensor([nan], dtype=torch.float64, device=device)
+                # Due to some deterministic issues with indices order in the GPU with repeated values
+                # 0 and 1 indices are both valid.
```
Comment (ngimel) [inline test/test_torch.py:8876]: wait, how can index 0 be valid?

File: test/test_torch.py [hunk 4]
```diff
@@ -8864,6 +8864,20 @@ def test_multidim(x, dim):
...
+                expected_indices = torch.tensor([1], dtype=torch.int64, device=device)
+                self.assertEqual(values, expected_values)
+                self.assertLessEqual(indices, expected_indices)
```
Comment (ngimel) [inline test/test_torch.py:8879]: I don't really like this test, maybe check that y[index] is nan?

File: test/test_torch.py (hunk 3 above)
Comment (emcastillo) [reply test/test_torch.py:8876]: sorry, I was still asleep 😅

File: test/test_torch.py (hunk 4 above)
Comment (emcastillo) [reply test/test_torch.py:8879]: sure!

**Comment by emcastillo** [comment]
Now should be ok 😅

**Review by ngimel — APPROVED** [review]
